In [1]:
import pickle 
import pandas as pd
df=pd.read_pickle('pandasFormat.pkl')

In [2]:
print(len(df.columns))
df.columns

28


Index(['innings', 'meta.data_version', 'meta.created', 'meta.revision',
       'info.dates', 'info.gender', 'info.match_type',
       'info.outcome.by.wickets', 'info.outcome.winner', 'info.overs',
       'info.player_of_match', 'info.teams', 'info.toss.decision',
       'info.toss.winner', 'info.umpires', 'info.venue', 'match_id',
       'info.city', 'info.outcome.by.runs', 'info.match_type_number',
       'info.neutral_venue', 'info.outcome.method', 'info.outcome.result',
       'info.outcome.eliminator', 'info.supersubs.New Zealand',
       'info.supersubs.South Africa', 'info.bowl_out',
       'info.outcome.bowl_out'],
      dtype='object')

In [3]:
df

innings  meta.data_version  \
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...                0.9   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...                0.9   
..                                                ...                ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...                0.9   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...                0.9   
0   [{'1st innings': {'team': 'Netherlands', 'deli...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...                0.9   

   meta.created  meta.revision    info.dates info.gender info.match_type  \
0    2017-02-18              2  [2017-02-17]        male             T20   
0    2017-02-19              2  [2017-02-19]        male             T20   
0    2017-02-23              1  [2017-02-22]        male             T20   
0    2016-09-12              1  [2016-09-05]        male             T20   
0    2016-06-19              1  [2016-06-18]        male             T20   
..          ...            ...           ...         ...             ...   
0    2016-03-05              2  [2016-03-04]        male             T20   
0    2016-03-08              1  [2016-03-06]        male             T20   
0    2016-02-03              1  [2016-02-03]        male             T20   
0    2016-09-12              1  [2016-09-06]        male             T20   
0    2016-09-12              1  [2016-09-09]        male             T20   

    info.outcome.by.wickets info.outcome.winner  info.overs  ...  \
0                       5.0           Sri Lanka          20  ...   
0                       2.0           Sri Lanka          20  ...   
0                       NaN           Australia          20  ...   
0                       NaN           Hong Kong          20  ...   
0                       NaN            Zimbabwe          20  ...   
..                      ...                 ...         ...  ...   
0                       6.0            Pakistan          20  ...   
0                       8.0               India          20  ...   
0                       NaN         Netherlands          20  ...   
0                       NaN           Australia          20  ...   
0                       4.0           Australia          20  ...   

   info.outcome.by.runs info.match_type_number info.neutral_venue  \
0                   NaN                    NaN                NaN   
0                   NaN                    NaN                NaN   
0                  41.0                    NaN                NaN   
0                  40.0                    NaN                NaN   
0                   2.0                    NaN                NaN   
..                  ...                    ...                ...   
0                   NaN                    NaN                1.0   
0                   NaN                    NaN                NaN   
0                  84.0                    NaN                NaN   
0                  85.0                    NaN                NaN   
0                   NaN                    NaN                NaN   

   info.outcome.method info.outcome.result info.outcome.eliminator  \
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
0                  NaN                 NaN                     NaN   
..                 ...                 ...                     ...   
0                  NaN                 NaN                     NaN   
0             

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1432 entries, 0 to 0
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   innings                      1432 non-null   object 
 1   meta.data_version            1432 non-null   float64
 2   meta.created                 1432 non-null   object 
 3   meta.revision                1432 non-null   int64  
 4   info.dates                   1432 non-null   object 
 5   info.gender                  1432 non-null   object 
 6   info.match_type              1432 non-null   object 
 7   info.outcome.by.wickets      662 non-null    float64
 8   info.outcome.winner          1386 non-null   object 
 9   info.overs                   1432 non-null   int64  
 10  info.player_of_match         1269 non-null   object 
 11  info.teams                   1432 non-null   object 
 12  info.toss.decision           1432 non-null   object 
 13  info.toss.winner     

In [5]:
df1=df.copy()

###  Below 1432 there is only one column that is necessary, "8-'info.outcome.winner'-1386 non-null-object ", so drop rest columns

In [6]:
df1 = df1.dropna(axis=1, thresh=1386)

In [7]:
df1.columns

Index(['innings', 'meta.data_version', 'meta.created', 'meta.revision',
       'info.dates', 'info.gender', 'info.match_type', 'info.outcome.winner',
       'info.overs', 'info.teams', 'info.toss.decision', 'info.toss.winner',
       'info.umpires', 'info.venue', 'match_id'],
      dtype='object')

In [8]:
df1

innings  meta.data_version  \
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...                0.9   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...                0.9   
..                                                ...                ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...                0.9   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...                0.9   
0   [{'1st innings': {'team': 'Netherlands', 'deli...                0.9   
0   [{'1st innings': {'team': 'Australia', 'delive...                0.9   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...                0.9   

   meta.created  meta.revision    info.dates info.gender info.match_type  \
0    2017-02-18              2  [2017-02-17]        male             T20   
0    2017-02-19              2  [2017-02-19]        male             T20   
0    2017-02-23              1  [2017-02-22]        male             T20   
0    2016-09-12              1  [2016-09-05]        male             T20   
0    2016-06-19              1  [2016-06-18]        male             T20   
..          ...            ...           ...         ...             ...   
0    2016-03-05              2  [2016-03-04]        male             T20   
0    2016-03-08              1  [2016-03-06]        male             T20   
0    2016-02-03              1  [2016-02-03]        male             T20   
0    2016-09-12              1  [2016-09-06]        male             T20   
0    2016-09-12              1  [2016-09-09]        male             T20   

   info.outcome.winner  info.overs                           info.teams  \
0            Sri Lanka          20               [Australia, Sri Lanka]   
0            Sri Lanka          20               [Australia, Sri Lanka]   
0            Australia          20               [Australia, Sri Lanka]   
0            Hong Kong          20                 [Ireland, Hong Kong]   
0             Zimbabwe          20                    [Zimbabwe, India]   
..                 ...         ...                                  ...   
0             Pakistan          20                [Pakistan, Sri Lanka]   
0                India          20                  [Bangladesh, India]   
0          Netherlands          20  [United Arab Emirates, Netherlands]   
0            Australia          20               [Sri Lanka, Australia]   
0            Australia          20               [Sri Lanka, Australia]   

   info.toss.decision      info.toss.winner  \
0               field             Sri Lanka   
0               field             Sri Lanka   
0               field             Sri Lanka   
0                 bat             Hong Kong   
0               field                 India   
..                ...                   ...   
0               field              Pakistan   
0               field                 India   
0               field  United Arab Emirates   
0               field             Sri Lanka   
0                 bat             Sri Lanka   

                          info.umpires  \
0               [MD Martell, P Wilson]   
0                [SD Fry, SJ Nogajski]   
0               [MD Martell, P Wilson]   
0                  [R Black, AJ Neill]   
0             [TJ Matibiri, RB Tiffin]   
..                                 ...   
0         [AK Chaudhary, Enamul Haque]   
0     [RSA Palliyaguruge, Shozab Raza]   
0           [CK Nandan, Sarika Prasad]   
0       [REJ Martinesz, RR Wimalasiri]   
0   [REJ Martinesz, RSA Palliyaguruge]   

                                 info.venue  match_id  
0                  Melbourne Cricket Ground         1  
0            Simonds Stadium, South Geelong         2  
0                             Adelaide Oval         3  
0         Bready Cricket 

### Drop other unnecessary columns

In [9]:
df1=df1.drop(columns=['meta.data_version','meta.created','meta.revision','info.umpires','match_id'])

In [10]:
df1

innings    info.dates  \
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
..                                                ...           ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
0   [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

   info.gender info.match_type info.outcome.winner  info.overs  \
0         male             T20           Sri Lanka          20   
0         male             T20           Sri Lanka          20   
0         male             T20           Australia          20   
0         male             T20           Hong Kong          20   
0         male             T20            Zimbabwe          20   
..         ...             ...                 ...         ...   
0         male             T20            Pakistan          20   
0         male             T20               India          20   
0         male             T20         Netherlands          20   
0         male             T20           Australia          20   
0         male             T20           Australia          20   

                             info.teams info.toss.decision  \
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                  [Ireland, Hong Kong]                bat   
0                     [Zimbabwe, India]              field   
..                                  ...                ...   
0                 [Pakistan, Sri Lanka]              field   
0                   [Bangladesh, India]              field   
0   [United Arab Emirates, Netherlands]              field   
0                [Sri Lanka, Australia]              field   
0                [Sri Lanka, Australia]                bat   

        info.toss.winner                               info.venue  
0              Sri Lanka                 Melbourne Cricket Ground  
0              Sri Lanka           Simonds Stadium, South Geelong  
0              Sri Lanka                            Adelaide Oval  
0              Hong Kong        Bready Cricket Club, Magheramason  
0                  India                       Harare Sports Club  
..                   ...                                      ...  
0               Pakistan            Shere Bangla National Stadium  
0                  India            Shere Bangla National Stadium  
0   United Arab Emirates                              ICC Academy  
0              Sri Lanka  Pallekele International Cricket Stadium  
0              Sri Lanka                      R Premadasa Stadium  

[1432 rows x 10 columns]

### Keep only those rows which is
###  'info.gender'=male    'info.match_type' = T20,      'info.overs'=20 , 

In [11]:
df1=df1[(df1['info.gender']=='male') & (df1['info.match_type']=='T20') & (df1['info.overs']==20 )]

In [12]:
df1

innings    info.dates  \
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
..                                                ...           ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
0   [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
0   [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

   info.gender info.match_type info.outcome.winner  info.overs  \
0         male             T20           Sri Lanka          20   
0         male             T20           Sri Lanka          20   
0         male             T20           Australia          20   
0         male             T20           Hong Kong          20   
0         male             T20            Zimbabwe          20   
..         ...             ...                 ...         ...   
0         male             T20            Pakistan          20   
0         male             T20               India          20   
0         male             T20         Netherlands          20   
0         male             T20           Australia          20   
0         male             T20           Australia          20   

                             info.teams info.toss.decision  \
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                  [Ireland, Hong Kong]                bat   
0                     [Zimbabwe, India]              field   
..                                  ...                ...   
0                 [Pakistan, Sri Lanka]              field   
0                   [Bangladesh, India]              field   
0   [United Arab Emirates, Netherlands]              field   
0                [Sri Lanka, Australia]              field   
0                [Sri Lanka, Australia]                bat   

        info.toss.winner                               info.venue  
0              Sri Lanka                 Melbourne Cricket Ground  
0              Sri Lanka           Simonds Stadium, South Geelong  
0              Sri Lanka                            Adelaide Oval  
0              Hong Kong        Bready Cricket Club, Magheramason  
0                  India                       Harare Sports Club  
..                   ...                                      ...  
0               Pakistan            Shere Bangla National Stadium  
0                  India            Shere Bangla National Stadium  
0   United Arab Emirates                              ICC Academy  
0              Sri Lanka  Pallekele International Cricket Stadium  
0              Sri Lanka                      R Premadasa Stadium  

[963 rows x 10 columns]

###  No. of rows reduced from 1432 to 963, now delete those 3 columns because it is constant

In [13]:
df1=df1.drop(columns=['info.gender','info.match_type','info.overs'])

In [14]:
df1=df1.drop(columns=['info.dates'])

In [15]:
df1

innings info.outcome.winner  \
0   [{'1st innings': {'team': 'Australia', 'delive...           Sri Lanka   
0   [{'1st innings': {'team': 'Australia', 'delive...           Sri Lanka   
0   [{'1st innings': {'team': 'Australia', 'delive...           Australia   
0   [{'1st innings': {'team': 'Hong Kong', 'delive...           Hong Kong   
0   [{'1st innings': {'team': 'Zimbabwe', 'deliver...            Zimbabwe   
..                                                ...                 ...   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...            Pakistan   
0   [{'1st innings': {'team': 'Bangladesh', 'deliv...               India   
0   [{'1st innings': {'team': 'Netherlands', 'deli...         Netherlands   
0   [{'1st innings': {'team': 'Australia', 'delive...           Australia   
0   [{'1st innings': {'team': 'Sri Lanka', 'delive...           Australia   

                             info.teams info.toss.decision  \
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                [Australia, Sri Lanka]              field   
0                  [Ireland, Hong Kong]                bat   
0                     [Zimbabwe, India]              field   
..                                  ...                ...   
0                 [Pakistan, Sri Lanka]              field   
0                   [Bangladesh, India]              field   
0   [United Arab Emirates, Netherlands]              field   
0                [Sri Lanka, Australia]              field   
0                [Sri Lanka, Australia]                bat   

        info.toss.winner                               info.venue  
0              Sri Lanka                 Melbourne Cricket Ground  
0              Sri Lanka           Simonds Stadium, South Geelong  
0              Sri Lanka                            Adelaide Oval  
0              Hong Kong        Bready Cricket Club, Magheramason  
0                  India                       Harare Sports Club  
..                   ...                                      ...  
0               Pakistan            Shere Bangla National Stadium  
0                  India            Shere Bangla National Stadium  
0   United Arab Emirates                              ICC Academy  
0              Sri Lanka  Pallekele International Cricket Stadium  
0              Sri Lanka                      R Premadasa Stadium  

[963 rows x 6 columns]

### Correct the indexing if its unordered, or zero for all rows


In [16]:
df1=df1.reset_index(drop=True)
df1

innings info.outcome.winner  \
0    [{'1st innings': {'team': 'Australia', 'delive...           Sri Lanka   
1    [{'1st innings': {'team': 'Australia', 'delive...           Sri Lanka   
2    [{'1st innings': {'team': 'Australia', 'delive...           Australia   
3    [{'1st innings': {'team': 'Hong Kong', 'delive...           Hong Kong   
4    [{'1st innings': {'team': 'Zimbabwe', 'deliver...            Zimbabwe   
..                                                 ...                 ...   
958  [{'1st innings': {'team': 'Sri Lanka', 'delive...            Pakistan   
959  [{'1st innings': {'team': 'Bangladesh', 'deliv...               India   
960  [{'1st innings': {'team': 'Netherlands', 'deli...         Netherlands   
961  [{'1st innings': {'team': 'Australia', 'delive...           Australia   
962  [{'1st innings': {'team': 'Sri Lanka', 'delive...           Australia   

                              info.teams info.toss.decision  \
0                 [Australia, Sri Lanka]              field   
1                 [Australia, Sri Lanka]              field   
2                 [Australia, Sri Lanka]              field   
3                   [Ireland, Hong Kong]                bat   
4                      [Zimbabwe, India]              field   
..                                   ...                ...   
958                [Pakistan, Sri Lanka]              field   
959                  [Bangladesh, India]              field   
960  [United Arab Emirates, Netherlands]              field   
961               [Sri Lanka, Australia]              field   
962               [Sri Lanka, Australia]                bat   

         info.toss.winner                               info.venue  
0               Sri Lanka                 Melbourne Cricket Ground  
1               Sri Lanka           Simonds Stadium, South Geelong  
2               Sri Lanka                            Adelaide Oval  
3               Hong Kong        Bready Cricket Club, Magheramason  
4                   India                       Harare Sports Club  
..                    ...                                      ...  
958              Pakistan            Shere Bangla National Stadium  
959                 India            Shere Bangla National Stadium  
960  United Arab Emirates                              ICC Academy  
961             Sri Lanka  Pallekele International Cricket Stadium  
962             Sri Lanka                      R Premadasa Stadium  

[963 rows x 6 columns]

### Upto now we have handled missing values, and we have kept only the required rows and columns and dropped rest.
### Now we will deal with 1st column 'innings'

In [17]:
df1.iloc[0,0]

[{'1st innings': {'team': 'Australia',
   'deliveries': [{0.1: {'batsman': 'AJ Finch',
      'bowler': 'SL Malinga',
      'non_striker': 'M Klinger',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.2: {'batsman': 'AJ Finch',
      'bowler': 'SL Malinga',
      'non_striker': 'M Klinger',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.3: {'batsman': 'AJ Finch',
      'bowler': 'SL Malinga',
      'non_striker': 'M Klinger',
      'runs': {'batsman': 1, 'extras': 0, 'total': 1}}},
    {0.4: {'batsman': 'M Klinger',
      'bowler': 'SL Malinga',
      'non_striker': 'AJ Finch',
      'runs': {'batsman': 2, 'extras': 0, 'total': 2}}},
    {0.5: {'batsman': 'M Klinger',
      'bowler': 'SL Malinga',
      'non_striker': 'AJ Finch',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.6: {'batsman': 'M Klinger',
      'bowler': 'SL Malinga',
      'non_striker': 'AJ Finch',
      'runs': {'batsman': 3, 'extras': 0, 'total': 3}}},
    {1.1: {'batsm

### We observed that each element in 1st column contains ball-by-ball info of both the inning, and it is very compact, nested dictionaries and lists


### Final model input:  
#### BattingTeam, BowlingTeam, Venue, Inning, Overs(more than 10), Score, Wickets, Runs in last 5 overs, Target(if 2nd inning)
### Output : Final score after 20 overs

In [18]:
df1.iloc[3,0][1]['2nd innings']['deliveries']

[{0.1: {'batsman': 'WTS Porterfield',
   'bowler': 'Tanwir Afzal',
   'non_striker': 'DA Rankin',
   'runs': {'batsman': 3, 'extras': 0, 'total': 3}}},
 {0.2: {'batsman': 'DA Rankin',
   'bowler': 'Tanwir Afzal',
   'non_striker': 'WTS Porterfield',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.3: {'batsman': 'DA Rankin',
   'bowler': 'Tanwir Afzal',
   'non_striker': 'WTS Porterfield',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.4: {'batsman': 'DA Rankin',
   'bowler': 'Tanwir Afzal',
   'non_striker': 'WTS Porterfield',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.5: {'batsman': 'DA Rankin',
   'bowler': 'Tanwir Afzal',
   'non_striker': 'WTS Porterfield',
   'runs': {'batsman': 1, 'extras': 0, 'total': 1}}},
 {0.6: {'batsman': 'WTS Porterfield',
   'bowler': 'Tanwir Afzal',
   'non_striker': 'DA Rankin',
   'runs': {'batsman': 1, 'extras': 0, 'total': 1}}},
 {1.1: {'batsman': 'WTS Porterfield',
   'bowler': 'Nadeem Ahmed',
   'non_striker': 'DA

In [19]:
dataset=[]

for i in range(len(df1)):
    try:
        dataset.append([])

        teams=df1.loc[i]['info.teams']
        toss_winner=df1.loc[i]['info.toss.winner']
        toss_choice=df1.loc[i]['info.toss.decision']
        batting_1st_inning = toss_winner if toss_choice == 'bat' else (set(teams) - set([toss_winner])).pop()
        bowling_1st_inning=(set(teams) - set([batting_1st_inning])).pop()
        # Ist inning
        df2=pd.DataFrame()
        score=0
        wickets=0
        deliveries=df1.iloc[i,0][0]['1st innings']['deliveries']
        for j in range(len(deliveries)):
            over=list(deliveries[j].keys())[0]
            run=deliveries[j][over]['runs']['total']
            score+=run
            wickets+=1 if 'wicket' in deliveries[j][over] else 0
            dic=pd.DataFrame({'over':[over],'score':[score],'wickets':[wickets]})
            df2=pd.concat([df2,dic],ignore_index=True)

        df2['batting']=batting_1st_inning
        df2['bowling']=bowling_1st_inning
        df2['inning']=1

        dataset[-1].append(df2)

        # IInd inning
        df2=pd.DataFrame()
        score=0
        wickets=0
        deliveries=df1.iloc[i,0][1]['2nd innings']['deliveries']
        for j in range(len(deliveries)):
            over=list(deliveries[j].keys())[0]
            run=deliveries[j][over]['runs']['total']
            score+=run
            wickets+=1 if 'wicket' in deliveries[j][over] else 0
            dic=pd.DataFrame({'over':[over],'score':[score],'wickets':[wickets]})
            df2=pd.concat([df2,dic],ignore_index=True)

        df2['batting']=bowling_1st_inning
        df2['bowling']=batting_1st_inning  
        df2['inning']=2
        dataset[-1].append(df2)
    except:
        print(i)
        dataset.pop()

65
111
136
211
226
305
431
652
698
817
899


#### In the above matches, there is some different data structure used, thats why its throwing error.
#### Below, we will found that few of them don't have 2nd inning played, We will discard that data.

In [20]:
df1.iloc[65,0]

[{'1st innings': {'team': 'New Zealand',
   'deliveries': [{0.1: {'non_striker': 'C Munro',
      'bowler': 'SS Cottrell',
      'runs': {'extras': 0, 'total': 0, 'batsman': 0},
      'batsman': 'MJ Guptill'}},
    {0.2: {'non_striker': 'C Munro',
      'bowler': 'SS Cottrell',
      'runs': {'extras': 0, 'total': 2, 'batsman': 2},
      'batsman': 'MJ Guptill'}},
    {0.3: {'non_striker': 'C Munro',
      'bowler': 'SS Cottrell',
      'runs': {'extras': 0, 'total': 0, 'batsman': 0},
      'batsman': 'MJ Guptill'}},
    {0.4: {'non_striker': 'C Munro',
      'bowler': 'SS Cottrell',
      'runs': {'extras': 1, 'total': 1, 'batsman': 0},
      'extras': {'wides': 1},
      'batsman': 'MJ Guptill'}},
    {0.5: {'non_striker': 'C Munro',
      'bowler': 'SS Cottrell',
      'runs': {'extras': 0, 'total': 0, 'batsman': 0},
      'batsman': 'MJ Guptill',
      'wicket': {'player_out': 'MJ Guptill',
       'fielders': ['CAK Walton'],
       'kind': 'caught'}}},
    {0.6: {'non_striker': 'C 

In [21]:
len(dataset)

952

In [22]:
df1.loc[10]

innings                [{'1st innings': {'team': 'South Africa', 'del...
info.outcome.winner                                         South Africa
info.teams                                   [New Zealand, South Africa]
info.toss.decision                                                 field
info.toss.winner                                             New Zealand
info.venue                                                     Eden Park
Name: 10, dtype: object

In [25]:
dataset[10][0]

over  score  wickets       batting      bowling  inning
0     0.1      0        0  South Africa  New Zealand       1
1     0.2      4        0  South Africa  New Zealand       1
2     0.3      4        0  South Africa  New Zealand       1
3     0.4      4        0  South Africa  New Zealand       1
4     0.5      5        0  South Africa  New Zealand       1
..    ...    ...      ...           ...          ...     ...
117  19.2    179        5  South Africa  New Zealand       1
118  19.3    179        5  South Africa  New Zealand       1
119  19.4    181        5  South Africa  New Zealand       1
120  19.5    181        6  South Africa  New Zealand       1
121  19.6    185        6  South Africa  New Zealand       1

[122 rows x 6 columns]

#### We have successfully expanded the first column, now we have only two column left-
#### 'runs in last 5 overs','final_score'  --------- venue and target(if its 2nd inning) will not be doing

In [61]:
for i in range(len(dataset)):
    print(i)
    for j in range(2):    ## j=0 or 1
        dataset[i][j]['Last5oversRuns']=-1
        for k in range(len(dataset[i][j])):
            over=dataset[i][j].loc[k]['over']
            if over>9.6:
                dataset[i][j].iloc[k,-1]=dataset[i][j].iloc[k,1] - dataset[i][j].iloc[k-30,1]
        dataset[i][j]['Final_score']=dataset[i][j].iloc[-1,1]

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [64]:
dataset[10][1]

over  score  wickets      batting       bowling  inning  Last5oversRuns  \
0    0.1      0        0  New Zealand  South Africa       2              -1   
1    0.2      0        0  New Zealand  South Africa       2              -1   
2    0.3      0        0  New Zealand  South Africa       2              -1   
3    0.4      0        0  New Zealand  South Africa       2              -1   
4    0.5      1        0  New Zealand  South Africa       2              -1   
..   ...    ...      ...          ...           ...     ...             ...   
86  14.1    106        8  New Zealand  South Africa       2              51   
87  14.2    106        8  New Zealand  South Africa       2              51   
88  14.3    106        9  New Zealand  South Africa       2              51   
89  14.4    107        9  New Zealand  South Africa       2              48   
90  14.5    107       10  New Zealand  South Africa       2              48   

    Final_score  
0           107  
1           107  
2           107  
3           107  
4           107  
..          ...  
86          107  
87          107  
88          107  
89          107  
90          107  

[91 rows x 8 columns]

#### Now we combine all the dataset into a single pandas dataframe

In [79]:
dataset2=pd.DataFrame()
for i in range(len(dataset)):
    for j in range(2):
        dataset2=pd.concat([dataset2,dataset[i][j]],ignore_index=True)
dataset2

over  score  wickets    batting    bowling  inning  Last5oversRuns  \
0        0.1      0        0  Australia  Sri Lanka       1              -1   
1        0.2      0        0  Australia  Sri Lanka       1              -1   
2        0.3      1        0  Australia  Sri Lanka       1              -1   
3        0.4      3        0  Australia  Sri Lanka       1              -1   
4        0.5      3        0  Australia  Sri Lanka       1              -1   
...      ...    ...      ...        ...        ...     ...             ...   
220053  17.1    121        6  Australia  Sri Lanka       2              18   
220054  17.2    124        6  Australia  Sri Lanka       2              21   
220055  17.3    124        6  Australia  Sri Lanka       2              21   
220056  17.4    124        6  Australia  Sri Lanka       2              21   
220057  17.5    130        6  Australia  Sri Lanka       2              26   

        Final_score  
0               168  
1               168  
2               168  
3               168  
4               168  
...             ...  
220053          130  
220054          130  
220055          130  
220056          130  
220057          130  

[220058 rows x 8 columns]

#### For training the model, we will choose only those rows after 10 overs, because we can't predict the final score if a match has not been played for 10 overs yet.

In [80]:
dataset2=dataset2[dataset2['Last5oversRuns'] != -1]

In [83]:
dataset2=dataset2.reset_index(drop=True)

In [84]:
dataset2

over  score  wickets    batting    bowling  inning  Last5oversRuns  \
0       10.1     75        0  Australia  Sri Lanka       1              30   
1       10.2     76        0  Australia  Sri Lanka       1              31   
2       10.3     76        1  Australia  Sri Lanka       1              31   
3       10.4     78        1  Australia  Sri Lanka       1              32   
4       10.5     78        1  Australia  Sri Lanka       1              31   
...      ...    ...      ...        ...        ...     ...             ...   
103180  17.1    121        6  Australia  Sri Lanka       2              18   
103181  17.2    124        6  Australia  Sri Lanka       2              21   
103182  17.3    124        6  Australia  Sri Lanka       2              21   
103183  17.4    124        6  Australia  Sri Lanka       2              21   
103184  17.5    130        6  Australia  Sri Lanka       2              26   

        Final_score  
0               168  
1               168  
2               168  
3               168  
4               168  
...             ...  
103180          130  
103181          130  
103182          130  
103183          130  
103184          130  

[103185 rows x 8 columns]

In [85]:
import pickle
dataset2.to_pickle('Dataset.pkl')